File: boxbot_gazebo/nodes/mimic.py
#!/usr/bin/env python

"""
  A script to support mimicking in Gazebo with the parallel grippers on the
  WidowX arms of Boxbot.

  Copyright 2018 University of Cincinnati
  All rights reserved. See LICENSE file at:
  https://github.com/MatthewVerbryke/inmoov-ros
  Additional copyright may be held by others, as reflected in the commit history.
  
"""


import sys
import thread

from std_msgs.msg import Float64
import rospy


class ParallelGripperMimic:
    """
    A ROS node that allows for mimicing joint commands on a parallel gripper
    (such as that on the WidowX) in Gazebo.
    
    TODO: TEST
    """
    
    def __init__(self):
        """
        
        """
        
        # Initialize node
        rospy.init_node("parallel_gripper_mimic")
        
        # Initialize cleanup
        rospy.on_shutdown(self.cleanup)
        
        # Get a lock
        self.lock = thread.allocate_lock()
        
        # Define topics 
        init_string = '/boxbot/'
        final_string = '/command'
        leaders = ["left_gripper_joint_controller", "right_gripper_joint_controller"]
        mimics = ["left_gripper_mimic_controller", "right_gripper_mimic_controller"]
        callbacks = [self.get_left_command, self.get_right_command]
        
        pubs = []
        subs = []
        
        # Setup publishers and subscribers
        for i in range(0,len(leaders)):
            hold_sub = rospy.Subscriber(init_string + leaders[i] + final_string, Float64, callbacks[i])
            hold_pub = rospy.Publisher(init_string + mimics[i] + final_string, Float64, queue_size=1)
            subs.append(hold_sub)
            pubs.append(hold_pub)
        
        # Wait for topics to start publishing before continuing
        for topic in subs:
            rospy.wait_for_message(init_string + leaders[i] + final_string, Float64)
            
        # Run main script
        self.main()
        
    def main(self):
        """
        Calculate and publish the nessecary commands for all mimiced joints.
        """
        
        while not rospy.is_shutdown():
            
            # Acquire a lock
            self.lock.aquire()
            
            try:
                
                # Update current commands
                current_goal = [self.left_command, self.right_command]
                
                # Publish the current commands to the parallel gripper
                for i in range(0,len(current_goal)):
                    pubs[i].publish(current_goal[i])
                    
            finally:
                
                # Release the lock
                self.lock.release()
                
    def get_left_command(self, msg):
        """
        Callback for the left gripper command.
        """
        self.left_command = msg.data
        
    def get_right_command(self, msg):
        """
        Callback for the right gripper command.
        """
        self.right_command = msg.data
        
    def cleanup(self):
        """
        Things to do on shutdown.
        """
        rospy.loginfo("Shutting down node 'parallel_gripper_mimic'")
        rospy.sleep(1)
        

if __name__ == "__main__":
    try:
        ParallelGripperMimic()
    except rospy.ROSInterruptException:
        pass

File: boxbot_gazebo/src/mimic.cpp
#include <string>

#include <ros/ros.h>
#include <std_msgs/Float64.h>

class Listener{
    
    public:
        double command = 10.0;
    
    void jointCommandCallback(const std_msgs::Float64& msg){
        command = msg.data;
    }
};
    
int main(int argc, char **argv){
    
    // Initialize node
    ros::init(argc, argv, "gripper_mimic");
    ros::NodeHandle nh("~");
    
    // Get ROS Parameters
    std::string side;
    nh.getParam("side", side);
    
    // Set the initial joint command
    double currentGoal = 0.0;
    
    // Loop Rate
    ros::Rate loop_rate(60);
    
    // Create topic names for node
    Listener listener;
    std::string mimicPubName = "/boxbot/" + side + "_gripper_mimic_controller/command";
    std::string mimicSubName = "/boxbot/" + side + "_gripper_joint_controller/command";
    
    // Setup mimic joint command publisher
    ros::Publisher controlPub = nh.advertise<std_msgs::Float64>(mimicPubName, 1);
    
    // Setup joint command subscriber
    ros::Subscriber jointSub = nh.subscribe(mimicSubName, 1, &Listener::jointCommandCallback, &listener);
    
    ros::spinOnce();
    
    // Main Loop
    while (ros::ok())
    {
        // Get the current command
        currentGoal = listener.command;
        
        // Construct and publish the mimic command
        std_msgs::Float64 mimic_msg;
        mimic_msg.data = currentGoal;
        controlPub.publish(mimic_msg);
        
        ros::spinOnce();
        loop_rate.sleep();
    }
}
